kernel/examples/inspect-table/src/main.rs

```diff
@@ -86,89 +78,77 @@ fn fields_in(field: &StructField) -> usize {
 }
 
 struct LogVisitor {
-    actions: Vec<Action>,
-    add_offset: usize,
-    remove_offset: usize,
-    protocol_offset: usize,
-    metadata_offset: usize,
-    set_transaction_offset: usize,
+    actions: Vec<(Action, usize)>,
+    offsets: HashMap<&'static str, (usize, usize)>,
     previous_rows_seen: usize,
 }
 
 impl LogVisitor {
-    fn new(log_schema: &SchemaRef) -> LogVisitor {
-        let mut offset = 0;
-        let mut add_offset = 0;
-        let mut remove_offset = 0;
-        let mut protocol_offset = 0;
-        let mut metadata_offset = 0;
-        let mut set_transaction_offset = 0;
+    fn new(log_schema: &'static SchemaRef) -> LogVisitor {
```

Comment:
Any reason for SchemaRef to be static?

Reply:
> there's a [Borrow<str> for String](https://doc.rust-lang.org/std/string/struct.String.html#impl-Borrow%3Cstr%3E-for-String) but no Borrow<&str> for String

wow you're right that is silly. TIL the difference between `str` and `&str` 👍 

> to avoid allocating a string for every map probe and/or adding a lifetime to the enclosing struct

that makes sense.

> can do that since it's a local trait

And we can't do this because Borrow is a foreign trait, and SchemaRef is a foreign struct (from inspect-table's perspective). While in the example, ColumnName was at least local.

Thanks @scovich!

Reply:
Update: Turns out if I pass e.g. `offsets[ADD_NAME]` instead of `offsets[&ADD_NAME]` then it works!